modules/relayers/src/lib.rs

```diff
@@ -241,6 +269,196 @@ pub mod pallet {
 				Ok(())
 			})
 		}
+
+		/// Register relayer intention to serve given messages lane.
+		///
+		/// Relayer that registers itself at given message lane gets a priority boost for his
+		/// message delivery transactions, **verified** at his slots (consecutive range of blocks).
+		///
+		/// Every lane registration requires additional stake. Relayer registration is considered
+		/// active while it is registered at least at one lane.
+		///
+		/// Relayer may request large reward here (using `expected_reward`), but in the end, the
+		/// reward amount is computed at the bridged (source chain). In the case if
+		/// [`DeliveryConfirmationPaymentsAdapter`] is used to register rewards, the maximal reward
+		/// per message is limited by the `MaxRewardPerMessage` parameter.
+		#[pallet::call_index(3)]
+		#[pallet::weight(Weight::zero())] // TODO
+		pub fn register_at_lane(
+			origin: OriginFor<T>,
+			lane: LaneId,
+			expected_reward: RewardAtSource,
+		) -> DispatchResult {
+			let relayer = ensure_signed(origin)?;
+
+			RegisteredRelayers::<T>::try_mutate(
+				&relayer.clone(),
+				move |maybe_registration| -> DispatchResult {
+					// we only allow registered relayers to have priority boosts
+					let mut registration = match maybe_registration.take() {
+						Some(registration) => registration,
+						None => fail!(Error::<T>::NotRegistered),
+					};
+
+					// cannot add another lane registration if relayer has already max allowed
+					// lane registrations
+					ensure!(
+						registration.register_at_lane(lane),
+						Error::<T>::FailedToRegisterAtLane
+					);
+
+					// TODO: ideally we shall use the candle auction for relayers selection (similar
```

Comment:
Those two TODOs is the biggest unresolved issue remaining in this PR. Candle auction requires additional runtime storage, so probably not the best solution here if we want the solution to be scalable. Thinking on how to alleviate that.

Reply:
TLDR: the problem is that our `next_set` has limited size (right now it is the same size as the `active_set`) and unhonest relayers may fill this set by claiming to work for the smallest possible reward (say, zero). Then, right before the bidding ends, they could deregister and epoch will start with an empty set, which leads us to the chaos we have now. Also they can hold their place for relayers, that are working for the largest possible reward. So even during "auction" there are relayers that are ready to work for 1DOT/message, they may be pushed out by spam relayers, working for zero and then, right before next epoch starts, they'll deregister and will be replaced by relayers working for 10DOT/message. So - last minute changes is a problem.

I have three solutions:
1) if `deregister_at_lane` is called by relayer, which has not relayed at least one message since he was included in the `next_set`, he is slashed with his lane stake. The only exception is when he is pushed from the relayers set by relayer with smaller `expected_reward`. This way spam relayers, who may fill the whole relayers set with reward = zero won't be able to deregister before next epoch start without losing their stake;
2) increase the size of the `next_set`. It would make it harder (more expensive) for spam relayers to fill the whole next set;
3) an alternate implementation of candle auction, based on relay chain randomness. Something like - during `on_initialize` we check if `random() < (epoch_end - current_block) / epoch_length` and then stop auction. So it likely ends near the epoch end, but noone knows the actual block. Relayers will likely to make actual stakes since beginning to actually stay in the final `next_set`. Relayers will know the stop in advance (once the relay block is ready), but since we'll do our computations from `on_initialize`, they won't be able to alter the set using transactions.

I think we need to implement all three options, but for this PR I'd start with the option (2) as the easiest one. Next one in the queue (in a follow up PR) is the (1) and the last is (3).

Reply:
Option (2) is implemented in https://github.com/paritytech/parity-bridges-common/pull/2605/commits/081523c762c6ff613fe454c81976892f56e07c78